lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -254,43 +254,49 @@ public void run() {
 	 * transport events/notifications
 	 ***************************************/
 
-	void onMessage(ProtocolMessage message) {
-		if(protocolListener != null)
-			protocolListener.onRawMessage(message);
-		switch(message.action) {
-		case HEARTBEAT:
-			onHeartbeat(message);
-			break;
-		case ERROR:
-			ErrorInfo reason = message.error;
-			if(reason == null)
-				Log.e(TAG, "onMessage(): ERROR message received (no error detail)");
-			else
-				Log.e(TAG, "onMessage(): ERROR message received; message = " + reason.message + "; code = " + reason.code);
+	void onMessage(ProtocolMessage message) throws AblyException {
```

Comment:
> Therefore it is possible that we will get deserialisation failures for messages (this is most likely) and if deserialisation succeeds it is possible we will get protocol errors.

Which-ever network do that kind of messing with data (not corrupting, but modifying) and saying `Deal with it` should be root of all evil :). If that is really the case, you are most certainly right to do extra pre-caution on client side.

I will start working on that starting tomorrow.